File: solutions/typescript/153-insert-delete-getrandom-o-1.ts
/**
 * Problem 153: Insert Delete GetRandom O(1) (LeetCode 380)
 * Difficulty: Med
 * Language: TypeScript
 */
class RandomizedSet {
  private map: Map<number, number>;
  private list: number[];
  constructor() { this.map = new Map(); this.list = []; }
  insert(val: number): boolean {
    if (this.map.has(val)) return false;
    this.map = new Map([...this.map, [val, this.list.length]]);
    this.list = [...this.list, val];
    return true;
  }
  remove(val: number): boolean {
    if (!this.map.has(val)) return false;
    const idx = this.map.get(val)!;
    const last = this.list[this.list.length - 1];
    const newList = [...this.list.slice(0, -1)];
    if (idx < newList.length) newList[idx] = last;
    const newMap = new Map([...this.map]);
    if (idx < newList.length) newMap.set(last, idx);
    newMap.delete(val);
    this.map = newMap; this.list = newList;
    return true;
  }
  getRandom(): number { return this.list[Math.floor(Math.random() * this.list.length)]; }
}

File: solutions/elixir/153-insert-delete-getrandom-o-1.ex
# Problem 153: Insert Delete GetRandom O(1) (LeetCode 380)
# Difficulty: Med
# Language: Elixir
# 
defmodule RandomizedSet do
  defstruct map: %{}, list: []
  def new(), do: %RandomizedSet{}
  def insert(%RandomizedSet{map: m, list: l} = s, val) do
    if Map.has_key?(m, val), do: {false, s},
    else: {true, %RandomizedSet{map: Map.put(m, val, length(l)), list: l ++ [val]}}
  end
  def remove(%RandomizedSet{map: m, list: l} = s, val) do
    if not Map.has_key?(m, val), do: {false, s},
    else: (
      idx = m[val]; last_val = List.last(l); new_list = List.delete_at(l, -1)
      new_map = m |> Map.delete(val)
      if idx < length(new_list) do
        {true, %RandomizedSet{map: Map.put(new_map, last_val, idx),
          list: List.replace_at(new_list, idx, last_val)}}
      else {true, %RandomizedSet{map: new_map, list: new_list}} end
    )
  end
end
